src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java

```diff
@@ -0,0 +1,185 @@
+package games.strategy.triplea.ui.mapdata;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.awt.Point;
+import java.awt.geom.Line2D;
+import java.awt.geom.Point2D;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import games.strategy.triplea.ui.MapPanel;
+
+public class RouteOptimizer {
+
+  public final boolean isInfiniteY;
+  public final boolean isInfiniteX;
+
+  private final MapPanel mapPanel;
+
+  private Point endPoint;
+  private int mapWidth;
+  private int mapHeight;
+
+  public RouteOptimizer(MapData mapData, MapPanel mapPanel) {
+    checkNotNull(mapData);
+    this.mapPanel = checkNotNull(mapPanel);
+    isInfiniteY = mapData.scrollWrapY();
+    isInfiniteX = mapData.scrollWrapX();
+  }
+
+  public Point[] getTranslatedRoute(Point... route) {
+    if (route == null) {
```

Comment:
may want to check if the route is empty as well.
Moving this check to the caller IMO is ideal, simply do not call the method with no-op params.